### PR TITLE
Add `Z3` option to CFG-based checkers

### PR DESCRIPTION
**Pull request by Raine-Yang-UofT**
## Proposed Changes

_(Describe your changes here. Also describe the motivation for your changes: what problem do they solve, or how do they improve the application or codebase? If this pull request fixes an open issue, [use a keyword to link this pull request to the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).)_

Add `z3` option to the following CFG -based checkers to ignore unfeasible nodes based on edge z3 constraints:

- `one-iteration-checker`
- `redundant-assignment-checker`
- `possibly-undefined-checker`
- `inconsistent-missing-return-checker` (only the checking of `missing-return-statement` message considers z3 constraints)

...

<details>
<summary>Screenshots of your changes (if applicable)</summary>

</details>

## Type of Change

_(Write an `X` or a brief description next to the type or types that best describe your changes.)_

| Type                                                                                    | Applies? |
| --------------------------------------------------------------------------------------- | -------- |
| 🚨 _Breaking change_ (fix or feature that would cause existing functionality to change) |          |
| ✨ _New feature_ (non-breaking change that adds functionality)                          |     X     |
| 🐛 _Bug fix_ (non-breaking change that fixes an issue)                                  |          |
| ♻️ _Refactoring_ (internal change to codebase, without changing functionality)          |          |
| 🚦 _Test update_ (change that _only_ adds or modifies tests)                            |          |
| 📚 _Documentation update_ (change that _only_ updates documentation)                    |          |
| 📦 _Dependency update_ (change that updates a dependency)                               |          |
| 🔧 _Internal_ (change that _only_ affects developers or continuous integration)         |          |

## Checklist

_(Complete each of the following items for your pull request. Indicate that you have completed an item by changing the `[ ]` into a `[x]` in the raw text, or by clicking on the checkbox in the rendered description on GitHub.)_

Before opening your pull request:

- [x] I have performed a self-review of my changes.
  - Check that all changed files included in this pull request are intentional changes.
  - Check that all changes are relevant to the purpose of this pull request, as described above.
- [x] I have added tests for my changes, if applicable.
  - This is **required** for all bug fixes and new features.
- [x] I have updated the project documentation, if applicable.
  - This is **required** for new features.
- [x] I have updated the project Changelog (this is required for all changes).
- [x] If this is my first contribution, I have added myself to the list of contributors.

After opening your pull request:

- [ ] I have verified that the pre-commit.ci checks have passed.
- [ ] I have verified that the CI tests have passed.
- [ ] I have reviewed the test coverage changes reported by Coveralls.
- [ ] I have [requested a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) from a project maintainer.

## Questions and Comments

_(Include any questions or comments you have regarding your changes.)_

### CODE REVIEW

File: python_ta/checkers/inconsistent_or_missing_returns_checker.py [hunk 1]
```diff
@@ -25,6 +25,17 @@ class InconsistentReturnChecker(BaseChecker):
...
         ),
     }
+    options = (
+        (
+            "z3",
+            {
+                "default": False,
+                "type": "yn",
+                "metavar": "<y or n>",
+                "help": "Only check for logically feasible nodes based on edge z3 constraints",
```
Comment (david-yz-liu) [inline python_ta/checkers/inconsistent_or_missing_returns_checker.py:35]: Reword this to "Use Z3 to restrict control flow checks to paths that are logically feasible" (and change throughout)

File: tests/test_custom_checkers/test_inconsistent_returns.py [hunk 2]
```diff
@@ -153,3 +154,84 @@ def func():
...
+        inconsistent_return_node, _ = mod.nodes_of_class(nodes.Return)
+
+        with self.assertAddsMessages(
+            pylint.testutils.MessageTest(
+                msg_id="inconsistent-returns",
+                node=inconsistent_return_node,
+            ),
+            ignore_position=True,
+        ):
+            self.checker.linter.config.z3 = True
```
Comment (david-yz-liu) [inline tests/test_custom_checkers/test_inconsistent_returns.py:182]: In general, rather than set this attribute manually, you should create a new *test suite class*, and then set the `z3: True` configuration value as a class variable in the test suite---see how other tests set config values. The new tests you're adding should be put under these new test suite classes that have `z3` enabled.

File: python_ta/checkers/inconsistent_or_missing_returns_checker.py [hunk 3]
```diff
@@ -71,6 +82,12 @@ def _check_return_statements(self, node) -> None:
         if has_return_annotation or has_return_value:
             for block, statement in return_statements.items():
                 if statement is None:
+                    # ignore unfeasible edges for missing return if z3 option is on
+                    if self.linter.config.z3 and all(
```
Comment (david-yz-liu) [inline python_ta/checkers/inconsistent_or_missing_returns_checker.py:86]: This is good, but let's do some simplification here. Please add a [property](https://docs.python.org/3/library/functions.html#property) to `CFGBlock` called `is_feasible`, and then use that here (and throughout your changes, where appropriate)

File: python_ta/cfg/graph.py [hunk 4]
```diff
@@ -166,32 +166,44 @@ def multiple_link_or_merge(self, source: CFGBlock, targets: List[CFGBlock]) -> N
...
                 self.link(source, target)
 
-    def get_blocks(self) -> Generator[CFGBlock, None, None]:
-        """Generate a sequence of all blocks in this graph."""
-        yield from self._get_blocks(self.start, set())
+    def get_blocks(self, only_feasible: bool = False) -> Generator[CFGBlock, None, None]:
+        """Generate a sequence of all blocks in this graph.
 
-    def _get_blocks(self, block: CFGBlock, visited: Set[int]) -> Generator[CFGBlock, None, None]:
+        When only_feasible is True, only generate blocks feasible from start based on edge z3 constraints
```
Comment (david-yz-liu) [inline python_ta/cfg/graph.py:172]: Add a period to the end of this sentence

File: python_ta/checkers/one_iteration_checker.py [hunk 5]
```diff
@@ -53,11 +64,20 @@ def _check_one_iteration(self, node: Union[nodes.For, nodes.While]) -> bool:
         if preds == []:
             return False
 
+        # the loop block is unfeasible
```
Comment (david-yz-liu) [inline python_ta/checkers/one_iteration_checker.py:67]: Overall, I think this algorithm is more complicated than it needs to be. First, you should be able to conditionally filter `preds` directly for only feasible edges after Line 62: `if self.linter.config.z3: preds = ...`

Second, as a general principle whenever you have a comment describing and if statement check, the comment should say something like "Check whether..." rather than stating the condition directly, as that is confusing to the reader.

File: tests/test_custom_checkers/test_inconsistent_returns.py [hunk 6]
```diff
@@ -153,3 +154,92 @@ def func():
...
         ):
             self.checker.visit_functiondef(func_node)
+
+
+class TestInconsistentReturnCheckerZ3Option(pylint.testutils.CheckerTestCase):
+    CHECKER_CLASS = InconsistentReturnChecker
+
+    def setup_method(self) -> None:
+        super().setup_method()
+        self.linter.config.z3 = True
```
Comment (david-yz-liu) [inline tests/test_custom_checkers/test_inconsistent_returns.py:164]: There's a more idiomatic way to pass in configuration options to `CheckerTestCase` subclasses; see other test suites for examples.

File: tests/test_custom_checkers/test_inconsistent_returns.py (hunk 2 above)
Comment (david-yz-liu) [reply tests/test_custom_checkers/test_inconsistent_returns.py:190]: This comment hasn't been resolved

File: tests/test_custom_checkers/test_missing_return_statements.py [hunk 7]
```diff
@@ -348,3 +349,92 @@ def division_missing_return(x, y) -> int:
...
+        mod = z3v.visitor.visit(astroid.parse(src))
+        mod.accept(CFGVisitor())
+        func_node = next(mod.nodes_of_class(nodes.FunctionDef))
+
+        with self.assertAddsMessages(
+            pylint.testutils.MessageTest(
+                msg_id="missing-return-statement",
+                node=func_node,
+                args="feasible_missing_return",
+                line=9,
```
Comment (david-yz-liu) [inline tests/test_custom_checkers/test_missing_return_statements.py:402]: use `ignore_position` for these tests (throughout)

File: tests/test_custom_checkers/test_possibly_undefined_checker.py [hunk 8]
```diff
@@ -469,3 +470,149 @@ def g(a: int) -> int:
...
+            return a
+        """
+        z3v = Z3Visitor()
+        mod = z3v.visitor.visit(astroid.parse(src))
+        mod.accept(CFGVisitor())
+        func_node = mod.body[0]
+        *_, name_node = mod.nodes_of_class(nodes.Name)
+
+        with self.assertNoMessages():
+            self.checker.visit_functiondef(func_node)
```
Comment (david-yz-liu) [inline tests/test_custom_checkers/test_possibly_undefined_checker.py:502]: Don't call two different `visit` methods here. If you need both to activate for the message to be triggered, visit the top-level `Module` node instead

**Review by david-yz-liu — COMMENTED** [review]
@Raine-Yang-UofT okay good work. I left some more comments. The `OneIterationChecker` is a bit more complicated and needs work. Please pull out those changes into a separate PR (incorporating my feedback), so that in this pull request there is only updates for the other three CFG checkers.

File: tests/test_custom_checkers/test_missing_return_statements.py (hunk 7 above)
Comment (Raine-Yang-UofT) [reply tests/test_custom_checkers/test_missing_return_statements.py:402]: Hi Professor, just to make sure that are we only using `ignore_position` for all z3-related test cases or all test cases. The z3-related tests should only focus on z3 functionality, so position of the error should not matter. However, I guess we do want to check for error position for other tests as we are using custom rendering for this checker.

Comment (david-yz-liu) [reply tests/test_custom_checkers/test_missing_return_statements.py]: Hi @Raine-Yang-UofT, yes that's a good point. For the checks that use custom rendering, it makes sense to keep the position information; but for the checks that don't, I want to standardize on _not_ including the location information in the test assertion, as that makes the tests more brittle in general.

**Review by david-yz-liu — COMMENTED** [review]
@Raine-Yang-UofT good work, I replied to one comment; other than that, please update from master and resolve the merge conflicts.

**Review by david-yz-liu — COMMENTED** [review]
@Raine-Yang-UofT the code is all good, but I realized that with this change, we should be ensuring the checks work with the standard `python_ta.check_all` (not just the individual checkers and tests).

To make this happen, I suspect you need to trigger the `Z3Visitor`, similar to how the `CFGVisitor` is triggered. But, for efficiency purposes please only trigger the `Z3Visitor` if the `z3` option is enabled.

You can test your work by calling `python_ta.check_all` manually on different examples, with the z3 option enabled.

File: python_ta/patches/transforms.py [hunk 9]
```diff
@@ -4,15 +4,18 @@
 from pylint.lint import PyLinter
 
 from ..cfg.visitor import CFGVisitor
+from ..transforms.z3_visitor import Z3Visitor
 
 
-def patch_ast_transforms():
+def patch_ast_transforms(linter_config):
```
Comment (david-yz-liu) [inline python_ta/patches/transforms.py:10]: Just pass in the `z3` value directly (a `bool`); please also add type annotations here

**Review by david-yz-liu — APPROVED** [review]
@Raine-Yang-UofT terrific job! This was a big one 👍
